Documentation/Evolution/DelimiterSyntax.md

```diff
@@ -0,0 +1,388 @@
+
+Due to the existing use of `/` in comment syntax and operators, there are some syntactic ambiguities to consider. While there are quite a few cases to consider, we do not feel that the impact of any individual case is sufficient to disqualify the syntax. Some of these ambiguities require a couple of source breaking language changes, and as such the `/.../` syntax requires upgrading to a new language mode in order to use.
+
+## Detailed design
+
+### Upgrade path
+
+Due to the source breaking changes needed for the `/.../` syntax, it will be introduced in Swift 6 mode. However, projects will be able to adopt it earlier by using the compiler flag `-enable-regex-literals`. Note this does not affect the extended literal `#/.../#`, which will be usable immediately.
+
+### Named typed captures
+
+Regex literals have their capture types statically determined by the capture groups present. This follows the same inference behavior as [the DSL][regex-dsl], and is explored in more detail in *[Strongly Typed Captures][strongly-typed-captures]*. One aspect of this that is currently unique to the literal is the ability to infer labeled tuple elements for named capture groups. For example:
+
+```swift
+func matchHexAssignment(_ input: String) -> (String, Int)? {
+  let regex = /(?<identifier>[[:alpha:]]\w*) = (?<hex>[0-9A-F]+)/
+  // regex: Regex<(Substring, identifier: Substring, hex: Substring)>
+  
+  guard let match = regex.matchWhole(input), 
+        let hex = Int(match.hex, radix: 16) 
+  else { return nil }
+  
+  return (match.identifier, hex)
+}
+```
+
+This allows the captures to be referenced as `match.identifier` and `match.hex`, in addition to numerically (like unnamed capture groups) as `match.1` and `match.2`. This label inference behavior is not available in the DSL, however users are able to [bind captures to named variables instead][dsl-captures].
+
+### Extended delimiters `#/.../#`, `##/.../##`
+
+Backslashes may be used to write forward slashes within the regex literal, e.g `/foo\/bar/`. However, this can be quite syntactically noisy and confusing. To avoid this, a regex literal may be surrounded by an arbitrary number of balanced octothorpes. This changes the delimiter of the literal, and therefore allows the use of forward slashes without escaping. For example:
+
+```swift
+let regex = #/usr/lib/modules/([^/]+)/vmlinuz/#
+// regex: Regex<(Substring, Substring)>
+```
+
+The number of pounds may be further increased to allow the use of e.g `/#` within the literal. This is similar in style to the raw string literal syntax introduced by [SE-0200], however it has a couple of key differences. Backslashes do not become literal characters. Additionally, a multi-line mode, where whitespace and line-ending comments are ignored, is entered when the opening delimiter is followed by a newline.
+
+```swift
+let regex = #/
+  /usr/lib/modules/ # Prefix
```

Comment:
Note this is a slightly different regex to the single-line version as it includes a `/` at the start, we probably ought to be consistent, what do you think? I mainly avoided it in the single-line version as GitHub syntax colors it as a comment 😬

Reply:
Any extra slashes are unintentional and the natural outcome of using slash as both part of a delimiter and an interior character.

It's interesting that `#//usr/local/#` and `#/user/local//#` would sensibly be lexed as comments by source tools. That's a downside of `/` we should mention too. I suppose the workaround is to escape the slash if it's at the very beginning or end, which is another little wrinkle that `#/.../#` doesn't alleviate.